### ISSUE TITLE
Router sharing

### ISSUE THREAD

**Issue by fizruk** [haskell-servant/servant#454]
The problem is stated in https://github.com/haskell-servant/servant/pull/451#issuecomment-207429027, but I'll reproduce it here.

Consider this simple API:

```haskell
type API = Capture "param" Int :> "something" :> Post '[] ()
```

For this API router would essentially be `Text -> Map Text RouterApplication`.

The question is whether this `Map` is shared for all `Text` arguments.
In general this is obviously not true for a function with this type, but I suspect that it is true for (existing) servant combinators (since `Map` does not depend on that `Text` value, or does it?).

Here're are some examples:

```haskell
-- | Always shares Map.
router1 :: Text -> Map Text RouterApplication
router1 = m
  where m = Map.singleton "book" app

-- | Never shares Map.
router3 :: Text -> Map Text RouterApplication
router3 s = Map.singleton s app
```

If sharing is not happening for `Router`, then it might affect performance.

/cc @kosmikus

**Comment by kosmikus** [comment]
Yes, thanks. I'm working on this.

**Comment by kosmikus** [comment]
Hopefully fixed in #457.

**Comment by kosmikus** [comment]
Closed via #457.
